### PR TITLE
Reduce memory allocation, remove some functions in Serializer.

### PR DESCRIPTION
**Pull request by seelabs**
@scottschurr @rec

### CODE REVIEW

File: src/ripple/protocol/Serializer.h [hunk 1]
```diff
@@ -298,7 +297,7 @@ class Serializer
     }
 
     // low-level VL length encode/decode functions
-    static Blob encodeVL (int length);
+    static int encodeVL (int length, std::array<std::uint8_t, 4>& result);
```
Comment (scottschurr) [inline src/ripple/protocol/Serializer.h]: Hi Scott, this looks like a great approach to not using a vector, especially since the (slightly odd) signature doesn't seem to go outside of `Serializer`.  Another possible signature would be to return a `std::pair<int, std::array<...>>`, but I'm not convinced that's any more intuitive.  The return value optimization would make the two signatures approximately the same cost in execution time.

It looks like you were able to make this change without changing any code outside of `Serializer.h` and `Serializer.cpp`.  Could this method be made `private`?  If it could, that suggests some of the other methods could be `private` as well.  You might consider expanding the scope of the pull request by doing an audit of `Serializer` and making as many methods `private` as possible.  Just a suggestion.

File: src/ripple/protocol/Serializer.h [hunk 2]
```diff
@@ -308,6 +307,8 @@ class Serializer
     static int decodeVLLength (int b1);
     static int decodeVLLength (int b1, int b2);
     static int decodeVLLength (int b1, int b2, int b3);
+private:
+    int addEncoded(int length);
```
Comment (scottschurr) [inline src/ripple/protocol/Serializer.h]: Hurrah for private methods!

File: src/ripple/shamap/impl/SHAMapTreeNode.cpp [hunk 3]
```diff
@@ -79,8 +79,8 @@ SHAMapTreeNode::SHAMapTreeNode (Blob const& rawNode,
 {
     if (format == snfWIRE)
     {
-        Serializer s (rawNode);
-        int type = s.removeLastByte ();
+        Serializer s (rawNode.begin (), rawNode.end () - 1);
```
Comment (scottschurr) [inline src/ripple/shamap/impl/SHAMapTreeNode.cpp]: Is there any chance at all that `rawNode` could be empty?  If it could happen this code might behave badly.  That might be worth guarding against.

Comment (seelabs) [inline src/ripple/shamap/impl/SHAMapTreeNode.cpp]: It's worth guarding against. The old code `asserted` and `removeLastByte` returned `-1`. I'll do the same here to keep consistent. Will do.

File: src/ripple/protocol/Serializer.h (hunk 1 above)
Comment (seelabs) [inline src/ripple/protocol/Serializer.h]: @scottschurr This code comes from a larger refactoring I was working on with serializer that's being sidelined for now. I'm keeping the branch around in case I ever get to return from it and those changes would go into that branch. I do agree with you, just not for this very tiny PR.

Comment (seelabs) [inline src/ripple/protocol/Serializer.h]: @scottschurr I prefer your std::pair return. Splitting the return type into an output param and returning the actual length is a little weird. Will change. Fixing the memory allocation was a win, I'm not concerned about the cost of return a four byte array (and RVO makes this a non-concern, as you say).

Comment (scottschurr) [inline src/ripple/protocol/Serializer.h]: I'm cool with not doing the privatizing audit.  It was definitely a scope creep kind of request.

File: src/ripple/protocol/Serializer.h (hunk 2 above)
Comment (rec) [inline src/ripple/protocol/Serializer.h]: Tiny quibble: consistent space before `(`.

File: src/ripple/protocol/impl/Serializer.cpp [hunk 4]
```diff
@@ -470,29 +445,35 @@ bool Serializer::getVLLength (int& length, int offset) const
     return true;
 }
 
-Blob Serializer::encodeVL (int length)
+int Serializer::addEncoded (int length)
 {
-    unsigned char lenBytes[4];
+    std::array<std::uint8_t, 4> result;
+    int l = encodeVL (length, result);
```
Comment (rec) [inline src/ripple/protocol/impl/Serializer.cpp]: `l` is generally a poor choice for variable names because of its similarity to `1` and `I`.

Why not just inline it into the next line?

File: src/ripple/protocol/impl/Serializer.cpp [hunk 5]
```diff
@@ -470,29 +445,35 @@ bool Serializer::getVLLength (int& length, int offset) const
...
     {
         length -= 193;
-        lenBytes[0] = 193 + static_cast<unsigned char> (length >> 8);
-        lenBytes[1] = static_cast<unsigned char> (length & 0xff);
-        return Blob (&lenBytes[0], &lenBytes[2]);
+        result[0] = 193 + static_cast<unsigned char> (length >> 8);
+        result[1] = static_cast<unsigned char> (length & 0xff);
+        return 2;
     }
     else if (length <= 918744)
```
Comment (rec) [inline src/ripple/protocol/impl/Serializer.cpp]: `918744`!

File: src/ripple/protocol/impl/Serializer.cpp [hunk 6]
```diff
@@ -470,29 +445,35 @@ bool Serializer::getVLLength (int& length, int offset) const
...
+                          std::array<std::uint8_t, 4>& result)
+{
     if (length <= 192)
     {
-        lenBytes[0] = static_cast<unsigned char> (length);
-        return Blob (&lenBytes[0], &lenBytes[1]);
+        result[0] = static_cast<unsigned char> (length);
+        return 1;
     }
     else if (length <= 12480)
```
Comment (rec) [inline src/ripple/protocol/impl/Serializer.cpp]: What are these magic numbers?  I'll look around to see...

Quibble: you don't need an else if the previous clause ends with a return.

Comment (seelabs) [inline src/ripple/protocol/impl/Serializer.cpp]: @rec Serializer could use some refactoring for sure, but I'd like to keep this PR very small.

**Comment by seelabs** [comment]
@rec @scottschurr Pushed changes to address your concerns.

File: src/ripple/protocol/Serializer.h [hunk 7]
```diff
@@ -297,17 +296,18 @@ class Serializer
...
     }
 
-    // low-level VL length encode/decode functions
-    static Blob encodeVL (int length);
+    static int decodeLengthLength (int b1);
+    static int decodeVLLength (int b1);
+    static int decodeVLLength (int b1, int b2);
+    static int decodeVLLength (int b1, int b2, int b3);
+private:
+    static std::pair<std::array<std::uint8_t, 4>, int> encodeVL (int length);
```
Comment (scottschurr) [inline src/ripple/protocol/Serializer.h]: Love the `private`.  Thanks.

File: src/ripple/shamap/impl/SHAMapTreeNode.cpp [hunk 8]
```diff
@@ -79,8 +79,10 @@ SHAMapTreeNode::SHAMapTreeNode (Blob const& rawNode,
 {
     if (format == snfWIRE)
     {
-        Serializer s (rawNode);
-        int type = s.removeLastByte ();
+        assert (!rawNode.empty ());
+        Serializer s (rawNode.begin (),
+                      rawNode.empty () ? rawNode.end () : rawNode.end () - 1);
+        int type = rawNode.empty () ? -1 : rawNode.back ();
```
Comment (scottschurr) [inline src/ripple/shamap/impl/SHAMapTreeNode.cpp]: Hmmm.  I like guarding against bad iterators.  That's a good thing.

What I'm noticing is that if `rawNode.empty()` is true we eventually get around to setting `type` to `-1` which causes a `throw std::runtime_error(...)`.  Since you're going to throw eventually anyway if `rawNode.empty()`, you could throw earlier in the code (replace the `assert` with a `throw`) and dispense with the ternary operators.  Seems like it might be a win.  Thoughts?

Comment (seelabs) [inline src/ripple/shamap/impl/SHAMapTreeNode.cpp]: That is a win. Will do.

**Comment by seelabs** [comment]
@rec @scottschurr Pushed changed to address your concerns.

**Comment by scottschurr** [comment]
LGTM
:+1:

**Comment by rec** [comment]
:+1:
